sisyphus/sisyphusd.go

```diff
@@ -0,0 +1,285 @@
+	pollGapDuration  time.Duration
+	numRerun         int
+	catchFlakesByRun bool
+	// optional
+	alert           *Alert
+	branchProtector *branchProtector
+	exitSignal      chan bool
+}
+
+// Config is the optional configuration to daemon
+type Config struct {
+	PollGapDuration  time.Duration
+	NumRerun         int
+	CatchFlakesByRun bool
+}
+
+// NewDaemon creates a Daemon
+func NewDaemon(protectedJobs []string,
+	prowProject, prowZone, gubernatorURL, gcsBucket string, cfg *Config) *Daemon {
+	var jobsWatched []*jobStatus
+	for _, jobName := range protectedJobs {
+		jobsWatched = append(jobsWatched, &jobStatus{
+			name:             jobName,
+			lastCheckedRunNo: 0,
+			pendingFirstRuns: make(map[int]time.Time),
+			rerunJobStats:    make(map[string]*FlakeStat),
+		})
+	}
+	daemon := Daemon{
+		jobsWatched:      jobsWatched,
+		prowAccessor:     NewProwAccessor(prowProject, prowZone, gubernatorURL, gcsBucket),
+		storage:          NewSisyphusStorage(),
+		pollGapDuration:  DefaultPollGapDuration,
+		numRerun:         DefaultNumRerun,
+		catchFlakesByRun: DefaultCatchFlakesByRun,
+		exitSignal:       make(chan bool, 1), // default channel never receives, hence not blocking
+	}
+	if cfg != nil {
+		if cfg.PollGapDuration.Nanoseconds() > 0 {
+			daemon.pollGapDuration = cfg.PollGapDuration
+		}
+		if cfg.NumRerun > 0 {
+			daemon.numRerun = cfg.NumRerun
+		}
+		daemon.catchFlakesByRun = cfg.CatchFlakesByRun
+	}
+	return &daemon
+}
+
+// GetConfig returns the Config in use by d
+func (d *Daemon) GetConfig() *Config {
+	return &Config{
+		PollGapDuration:  d.pollGapDuration,
+		NumRerun:         d.numRerun,
+		CatchFlakesByRun: d.catchFlakesByRun,
+	}
+}
+
+// SetAlert activates email alerts to receiverAddr when jobs failed
+func (d *Daemon) SetAlert(gmailAppPass, identity, senderAddr, receiverAddr string,
+	alertConfig *AlertConfig) error {
+	var err error
+	d.alert, err = NewAlert(gmailAppPass, identity, senderAddr, receiverAddr, alertConfig)
+	return err
+}
+
+// SetProtectedBranch disables auto merging PRs on protected branch if jobs failed
+func (d *Daemon) SetProtectedBranch(owner, token, repo, branch string) {
+	d.branchProtector = newBranchProtector(owner, token, repo, branch)
+}
+
+// SetExitSignal sets a channel on daemon
+func (d *Daemon) SetExitSignal(channel chan bool) {
+	d.exitSignal = channel
+}
+
+// Start activates the daemon to start polling Prow results
+func (d *Daemon) Start() {
+	for {
+		newFailures := d.checkOnJobsWatched()
+		if d.alert != nil {
+			if newFailures != nil {
+				log.Printf("%d tests failed in last circle", len(newFailures))
+				if err := d.alert.Send(d.formatMessage(newFailures)); err != nil {
+					log.Printf("Unable to send alerts: %v", err)
+				}
+			} else {
+				log.Printf("No new tests failed in last circle.")
+			}
+		}
+		if d.branchProtector != nil {
+			d.branchProtector.process(newFailures)
+		}
+		log.Printf("Sleeping for %d", d.pollGapDuration)
+		select {
+		case <-d.exitSignal:
```

Comment:
probably pass a context to start instead. It will also ease testing as you can pass a timeout context.
Also when doing a for loop like that it might be good to call a private startOne() method that does one iteration only, that way you check the expected output of one iteration in a test.

Reply:
Done